File: compiler/internal/vm/core/SandboxIR/Module.cpp
//===- Module.cpp ---------------------------------------------------------===//
//
// Copyright (c) NeXTHub Corporation. All Rights Reserved.
// DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
//
// Author: Tunjay Akbarli
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at:
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Please contact NeXTHub Corporation, 651 N Broad St, Suite 201,
// Middletown, DE 19709, New Castle County, USA.
//
//===----------------------------------------------------------------------===//

#include "vm/core/SandboxIR/Module.h"
#include "vm/core/SandboxIR/Constant.h"
#include "vm/core/SandboxIR/Context.h"
#include "vm/core/SandboxIR/Function.h"
#include "vm/core/SandboxIR/Value.h"

using namespace vm::core::sandboxir;

Function *Module::getFunction(StringRef Name) const {
  toolchain::Function *LLVMF = LLVMM.getFunction(Name);
  return cast_or_null<Function>(Ctx.getValue(LLVMF));
}

GlobalVariable *Module::getGlobalVariable(StringRef Name,
                                          bool AllowInternal) const {
  return cast_or_null<GlobalVariable>(
      Ctx.getValue(LLVMM.getGlobalVariable(Name, AllowInternal)));
}

GlobalAlias *Module::getNamedAlias(StringRef Name) const {
  return cast_or_null<GlobalAlias>(Ctx.getValue(LLVMM.getNamedAlias(Name)));
}

GlobalIFunc *Module::getNamedIFunc(StringRef Name) const {
  return cast_or_null<GlobalIFunc>(Ctx.getValue(LLVMM.getNamedIFunc(Name)));
}

#ifndef NDEBUG
void Module::dumpOS(raw_ostream &OS) const { OS << LLVMM; }

void Module::dump() const {
  dumpOS(dbgs());
  dbgs() << "\n";
}
#endif // NDEBUG

File: compiler/internal/vm/static_core/plugins/ets/tests/interop_js/tests/class_operations/module.js
/**
 * Copyright (c) 2021-2025 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
'use strict';

export function evaluateNumber(v0, v1) {
	return v0 + v1;
}

export function doNothing() {
	print('Hello from empty function');
}

export function evaluateObject(obj) {
	return new ExampleClass(obj.v0 + obj.v1, obj.v0 + obj.v1);
}

export function evaluateArray(arr, size) {
	let result = [];
	for (let i = 0; i < size; i++) {
		result[i] = arr[i] + i * i;
	}
	return result;
}

export class ExampleClass {
	constructor(v0, v1) {
		this.v0 = v0;
		this.v1 = v1;
	}

	static evaluateNumber(v2, v3) {
		return v2 + v3;
	}

	static evaluateArray(arr, size) {
		let result = [];
		for (let i = 0; i < size; i++) {
			result[i] = arr[i] + i * i;
		}
		return result;
	}

	instanceEvaluateNumber() {
		// without 'this' also not working
		return this.v0 + this.v1;
	}

	evaluateObject(obj) {
		return new ExampleClass(obj.v0 + this.v1, this.v0 + obj.v1);
	}

	getV0() {
		return this.v0;
	}

	getV1() {
		return this.v1;
	}

	static emptyMethod() {
		print('Hello from empty method');
	}
}

export class ClassWithEmptyConstructor {
	constructor() {
		this.v0 = 42;
		this.v1 = 42;
	}

	verifyProperties(v0Expected, v1Expected) {
		if (this.v0 === v0Expected && this.v1 === v1Expected) {
			return 0;
		} else {
			return 1;
		}
	}

	getV0() {
		return this.v0;
	}

	getV1() {
		return this.v1;
	}
}

export class Kitten {
    constructor(id, name) {
      this.id = id;
      this.name = name;
      this.male = false;
    }
  }

export const createKitten = (id, name) => new Kitten(id, name);

export const MyNamespace = {
    Kitten,
    createKitten
};
